### ISSUE TITLE
When using tomcat-jdbc in its default configuration, DataSourceHealthIndicator doesn't recover when MySQL is restarted

### ISSUE THREAD

**Issue by wilkinsona** [spring-projects/spring-boot#1303]
Start MySQL and check the health:

```
{"status":"UP","database":"MySQL","hello":1}
```

Stop MySQL and check health again:

```
{
"database": "MySQL", "error": "org.springframework.dao.RecoverableDataAccessException: StatementCallback; SQL [SELECT 1]; Communications link failure\n\nThe last packet successfully received from the server was 47,301 milliseconds ago.  The last packet sent successfully to the server was 6 milliseconds ago.; nested exception is com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure\n\nThe last packet successfully received from the server was 47,301 milliseconds ago.  The last packet sent successfully to the server was 6 milliseconds ago.", "status": "DOWN"
}
```

Restart MySQL and check health:

```
{"status":"DOWN","error":"org.springframework.dao.DataAccessResourceFailureException: ConnectionCallback; SQL []; No operations allowed after connection closed.; nested exception is com.mysql.jdbc.exceptions.jdbc4.MySQLNonTransientConnectionException: No operations allowed after connection closed."}
```

It looks like it's not letting go of the broken connection. This was with MySQL 5.6.19 and using the `tomcat-jdbc` connection pool with minimal configuration, i.e. just `spring.datasource.url`.

**Comment by philwebb** [comment]
I think we should probably push this back to 1.3. @wilkinsona feel free to move back to 1.2.x if you disagree.

**Comment by wilkinsona** [comment]
Rather than focussing on the health indicator, we've decided that we should be more opinionated but the defaults for the auto-configured data source. `tomcat-jdbc` should, be default, be configured to call `Connection.isValid`. This will align its default behaviour with Hikari.

**Comment by snicoll** [comment]
This one is a dup of #4906

**Comment by snicoll** [comment]
I just tried with `1.4.0.BUILD-SNAPSHOT` and this is indeed fixed thanks to #4906
